devito/finite_differences/derivative.py

```diff
@@ -396,15 +396,25 @@ def _eval_fd(self, expr, **kwargs):
         Evaluate the finite-difference approximation of the Derivative.
         Evaluation is carried out via the following three steps:
 
-        - 1: Evaluate derivatives within the expression. For example given
+        - 1: Interpolate non-derivative shifts.
+            I.e u[x, y].dx(x0={y: y + h_y/2}) requires to interpolate `u` in `y`
+        - 2: Evaluate derivatives within the expression. For example given
             `f.dx * g`, `f.dx` will be evaluated first.
-        - 2: Evaluate the finite difference for the (new) expression.
+        - 3: Evaluate the finite difference for the (new) expression.
              This in turn is a two-step procedure, for Functions that may
              may need to be evaluated at a different point due to e.g. a
              shited derivative.
-        - 3: Apply substitutions.
+        - 4: Apply substitutions.
         """
-        # Step 1: Evaluate derivatives within expression
+        # Step 1: Evaluate non-derivative x0. We currently enforce a simple 2nd order
+        # interpolation to avoid very expensive finite differences on top of it.
+        x0_interp = {k: v for k, v in self.x0.items() if k not in self.dims
```

Comment:
this is hardly readable, I'd turn it into a plain for loop and separate statements
also, `k -> d`